File: nim/rc4.nim
import std/monotimes
import times

type
  RC4 = object
    s: array[256, uint16]
    i: uint16
    j: uint16

proc init(rc4: var RC4, key: string) =
  var
    data: array[256, uint16]
    j: uint16

  for i in 0..<256:
    data[i] = uint16(i)

  for i in 0..<256:
    let keybyte = uint16(ord(key[i mod key.len]))
    j = (j + data[i] + keybyte) mod 256

    var tmp = data[i]
    data[i] = data[j]
    data[j] = tmp

    rc4.s = data

proc get_byte(rc4: var RC4): uint16 =
  rc4.i = (rc4.i + 1) mod 256
  rc4.j = (rc4.j + rc4.s[rc4.i]) mod 256

  let tmp = rc4.s[rc4.i]
  rc4.s[rc4.i] = rc4.s[rc4.j]
  rc4.s[rc4.j] = tmp

  let idx = (rc4.s[rc4.i] + rc4.s[rc4.j]) mod 256
  return rc4.s[idx]

proc bench() =
  let start = getMonoTime()
  const
    key = "Keyfobsrulethebestofall"

  var
    rc4: RC4
    sum: uint64

  rc4.init(key)

  for i in 0..<250000000:
    let byte: uint16 = rc4.get_byte()
    sum += uint64(byte)

  let delta = (getMonoTime() - start).inMilliseconds()

  echo "|nim|rc4|", $sum, "|", delta, "|"

when isMainModule:
  bench()

File: dart/rc4.dart
import 'dart:io';
import 'dart:typed_data';
import 'dart:convert';

class RC4 {
  Uint8List s = Uint8List(256);
  int i = 0;
  int j = 0;

  RC4(Uint8List key) {
    for (var i = 0; i < 256; i++) {
      s[i] = i;
    }

    var j = 0;
    for (var i = 0; i < 256; i++) {
      j = (j + s[i] + key[i % key.length]) % 256;

      var tmp = s[j];
      s[j] = s[i];
      s[i] = tmp;
    }
  }

  int get_byte() {
    i = (i + 1) % 256;
    j = (j + s[i]) % 256;

    var tmp = s[j];
    s[j] = s[i];
    s[i] = tmp;

    var idx = (s[i] + s[j]) % 256;
    return s[idx];
  }
}

void main() {
  var sw = Stopwatch()..start();
  var key = Uint8List.fromList(ascii.encode("Keyfobsrulethebestofall"));
  var rc4 = RC4(key);

  var sum = 0;
  for (var i = 0; i < 250000000; i++) {
    sum += rc4.get_byte();
  }

  var timedelta = sw.elapsed.inMilliseconds;
  stdout.write("|dart|rc4|$sum|$timedelta|");
}
